### PR TITLE
[ETCM-927] enhance BlockchainTests/ValidBlocks/bcMultiChainTest/Chain…

### PR DESCRIPTION
**Pull request by strauss-m**
# Description

Enhance the test ChainAtoChainB_difficultyB.
Currently, the test can't completely pass due to a missing feature: transactions are not executed on blocks that are not part of the canonical blockchain.

# Proposed Solution

- a block insertion on an alternate chain is now valid and queryable by hash
- fields s, r and v in EthTransactionResponse are now pruned from leading 0.

# Testing

The test `BlockchainTests/ValidBlocks/bcMultiChainTest/ChainAtoChainB_difficultyB` fails later, on the storage check, instead at the first insertion of a block on the non-canonical (best) chain.

```
Request: {"jsonrpc":"2.0","method":"debug_storageRangeAt","params":["3", 2, "0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b", "0x0000000000000000000000000000000000000000000000000000000000000000", 20],"id":26}
Reply: {"jsonrpc":"2.0","result":{"complete":true,"storage":{}},"id":26}
Error: Compare States: Missing expected address: '0x095e7baea6a6c7c4c2dfeb977efac326af552d87' (bcMultiChainTest/ChainAtoChainB_difficultyB_Istanbul, fork: Istanbul, block: 5)
Error: CompareStates failed with errors: MissingExpectedAccount (bcMultiChainTest/ChainAtoChainB_difficultyB_Istanbul, fork: Istanbul, block: 5)
```

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala [hunk 1]
```diff
@@ -76,12 +76,28 @@ class BlockQueue(blockchain: Blockchain, val maxQueuedBlockNumberAhead: Int, val
     }
   }
 
+  /**
+    * Remove all blocks currently queued
+    */
+  def clear(): Unit = {
+    blocks.clear()
+    parentToChildren.clear()
+  }
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala]: This seems to be unused as we rebuild a new BlockQueue when needed.

File: src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala [hunk 2]
```diff
@@ -42,9 +42,18 @@ trait TestModeServiceBuilder extends LedgerBuilder {
...
+      * Returns the chain weight if the block represented by its hash is either stored in the blockchain or enqueued
+      *
+      * @param hash the hash of the block to retrieve the chain weight
+      * @return
+      */
+    override def getChainWeightByHash(hash: ByteString): Option[ChainWeight] = testLedger.getChainWeightByHash(hash)
   }
 
   override lazy val ledger: Ledger = new TestLedgerProxy
+//  override lazy val ledger: Ledger = testModeComponentsProvider.ledger(blockchainConfig, SealEngineType.NoReward)
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala]: we can remove this :-)

File: src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala [hunk 3]
```diff
@@ -42,9 +42,18 @@ trait TestModeServiceBuilder extends LedgerBuilder {
     ): Task[BlockImportResult] = testLedger.importBlock(block)
     override def resolveBranch(headers: NonEmptyList[BlockHeader]): BranchResolutionResult =
       testLedger.resolveBranch(headers)
+
+    /**
+      * Returns the chain weight if the block represented by its hash is either stored in the blockchain or enqueued
+      *
+      * @param hash the hash of the block to retrieve the chain weight
+      * @return
+      */
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala]: no need to repeat the doc I think

File: src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala [hunk 4]
```diff
@@ -19,14 +21,42 @@ class TestModeComponentsProvider(
...
 
-  def ledger(blockchainConfig: BlockchainConfig, sealEngine: SealEngineType): Ledger =
-    new LedgerImpl(
-      blockchain,
-      blockchainConfig,
-      syncConfig,
-      consensus(blockchainConfig, sealEngine),
-      validationExecutionContext
+  private var cache = HashMap.empty[(BlockchainConfig, SealEngineType), Ledger]
+  private var blockQueue = BlockQueue(blockchain, syncConfig)
```
Comment (AurelienRichez) [inline src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala]: Why do we need a cache ? We would like to minimize state if possible. I think there is always only one ledger at a given time anyway during tests right ?

I guess we want to keep the same BlockQueue between the calls. Initially I tried to put all the state in the same place in `TestService`. It's not really a perfect place but it avoids scattering the state everywhere and be able to simplify it when possible. To be consistent we could inject `blockQueue` in `def ledger`, or put the current `Ledger`  in `TestService` ?

**Review by AurelienRichez — COMMENTED** [review]
A few remarks. I don't know how it will play out with @jvdp effort to remove Ledger however.

Anyway we have more ETS tests passing so 👍 .

File: src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala (hunk 4 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala]: In my test case, I got two different ledger instances due to different configs.
But maybe I should investigate deeper into this and make sure that for a given test, only one blockchain config is used, do you think it's a worthwhile effort ?
If there is only one config, one can drop the cache.

File: src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala (hunk 1 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala]: true.
It was belonging to a fix change that has been removed. I'll clean up.

File: src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala (hunk 3 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala]: removed

File: src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala (hunk 2 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestModeServiceBuilder.scala]: removed

**Comment by jvdp** [comment]
> I don't know how it will play out with @jvdp effort to remove Ledger however.

It'll be interesting but I'm not afraid!

File: src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala [hunk 5]
```diff
@@ -30,9 +31,18 @@ class TestEthBlockServiceWrapper(blockchain: Blockchain, ledger: Ledger, consens
     .getByBlockHash(request)
     .map(
       _.map(blockByBlockResponse => {
-        val fullBlock = blockchain.getBlockByNumber(blockByBlockResponse.blockResponse.get.number).get
-        BlockByBlockHashResponse(blockByBlockResponse.blockResponse.map(response => toEthResponse(fullBlock, response)))
-      })
+        blockByBlockResponse.blockResponse
+          .toRight("missing block response")
```
Comment (leo-bogastry) [inline src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala]: Can you add some more specific data to error?

Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala]: I've added a little bit more context with regard to the error messages

File: src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala [hunk 6]
```diff
@@ -32,9 +33,19 @@ class TestEthBlockServiceWrapper(
...
+          .flatMap(baseBlockResponse => baseBlockResponse.hash.toRight(s"missing hash for block $baseBlockResponse"))
+          .flatMap(hash => ledger.getBlockByHash(hash).toRight(s"Ledger: unable to find block for hash=${hash.toHex}"))
+          .map(fullBlock =>
+            BlockByBlockHashResponse(
+              blockByBlockResponse.blockResponse.map(response => toEthResponse(fullBlock, response))
+            )
+          )
+          .left
+          .map(errorMessage => JsonRpcError.LogicError(errorMessage))
+      }).flatten
```
Comment (leo-bogastry) [inline src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala]: You can make a few simplifications here:
- move the import to the top of the file, please
- remove the final `.flatten` and use `flatMap` instead
- you can remove `{` after `=>`, by replacing the `(` before `blockByBlockResponse` by `{`
In summary from:
`_.map(blockByBlockResponse => {`  to `_.flatMap { blockByBlockResponse =>`

File: src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala (hunk 4 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala]: Were you able to investigate this?

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthBlocksService.scala [hunk 7]
```diff
@@ -74,8 +74,8 @@ class EthBlocksService(
     */
   def getByBlockHash(request: BlockByBlockHashRequest): ServiceResponse[BlockByBlockHashResponse] = Task {
     val BlockByBlockHashRequest(blockHash, fullTxs) = request
-    val blockOpt = blockchainReader.getBlockByHash(blockHash)
-    val weight = blockchain.getChainWeightByHash(blockHash)
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/jsonrpc/EthBlocksService.scala:75]: Are these two changes required by ETS? I wonder what would be a convenient way to refactor this with ledger gone...

File: src/main/scala/io/iohk/ethereum/jsonrpc/TestService.scala [hunk 8]
```diff
@@ -157,6 +161,9 @@ class TestService(
 
     // remove current genesis (Try because it may not exist)
     Try(blockchain.removeBlock(blockchain.genesisHeader.hash, withState = false))
+    // TODO clear the storage ? When relaunching some tests on the same running test mantis client,
+    // we end up with duplicate blocks because they are still present in the storage layer
+    // for example: bcMultiChainTest/ChainAtoChainB_BlockHash_Istanbul
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/jsonrpc/TestService.scala:167]: How about when `setChainParams` is called?

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala [hunk 9]
```diff
@@ -29,6 +29,13 @@ trait Ledger {
     */
   def getBlockByHash(hash: ByteString): Option[Block]
 
+  /**
+    * Returns the chain weight if the block represented by its hash is either stored in the blockchain or enqueued
+    * @param hash the hash of the block to retrieve the chain weight
+    * @return the chain weight if the block represented by its hash is either stored in the blockchain or enqueued
+    */
+  def getChainWeightByHash(hash: ByteString): Option[ChainWeight]
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: Yeah please don't add more things to ledger 😅

File: src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala [hunk 10]
```diff
@@ -31,10 +33,19 @@ class TestEthBlockServiceWrapper(
...
+          .flatMap(baseBlockResponse => baseBlockResponse.hash.toRight(s"missing hash for block $baseBlockResponse"))
+          .flatMap(hash => ledger.getBlockByHash(hash).toRight(s"Ledger: unable to find block for hash=${hash.toHex}"))
+          .map(fullBlock =>
+            BlockByBlockHashResponse(
+              blockByBlockResponse.blockResponse.map(response => toEthResponse(fullBlock, response))
+            )
+          )
+          .left
+          .map(errorMessage => JsonRpcError.LogicError(errorMessage))
+      }
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala:60]: This is a bit hard to follow, might be nicer as `for`-comprehension?

File: src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala (hunk 4 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestModeComponentsProvider.scala]: No, I didn't push further the idea. I don't know if the suite is actually using different config depending on tests, which would prevent a unique one (initial setup + each test config) convergence, and require the current multi-config cache.

File: src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala (hunk 6 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/testmode/TestEthBlockServiceWrapper.scala]: thanks, update done

**Review by leo-bogastry — APPROVED** [review]

File: src/main/scala/io/iohk/ethereum/jsonrpc/EthBlocksService.scala (hunk 7 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/jsonrpc/EthBlocksService.scala:75]: The underlying need was to be able to get a block by hash, no matter if it's on the canonical chain or an alternate one. The solution was to delegate the work to the ledger, which implements the getBlockByHash exactly as it (first look in the canonical blockchain, then into the BlockQueue). The getChainWeightByHash was added to match the getBlockByHash behaviour.

The feature (being able to query a block by its hash, no matter its location) is required by the ETS. After a successful insert, retesteth queries for the insert block by its hash to do some consistency check.
It is not restricted to the ledger, so if any other component provides this after the refactor, let's use the new one.

File: src/main/scala/io/iohk/ethereum/jsonrpc/TestService.scala (hunk 8 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/jsonrpc/TestService.scala:167]: do you mean put the cleaning into the setChainParams call, or into the setChainParams caller ?

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala (hunk 9 above)
Comment (strauss-m) [reply src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: It is linked to the Ledger.getBlockByHash call.
According to their usage in EthBlocksService, getBlockByHash() and getChainWeightByHash() should both be available on whatever component provides them.
The refactor should be straightforward, wherever the getBlockByHash() method ends up, you can put the getChainWeightByHash() one too. The underlying behaviour is also similar to getBlockByHash: first query the (canonical) BlockChain, then the BlockQueue if not found.

**Review by AurelienRichez — APPROVED** [review]
